Content.Shared/Movement/Systems/SlowContactsSystem.cs

```diff
@@ -80,30 +66,32 @@ private void MovementSpeedCheck(EntityUid uid, SlowedByContactComponent componen
 
             walkSpeed = Math.Min(walkSpeed, slowContactsComponent.WalkSpeedModifier);
             sprintSpeed = Math.Min(sprintSpeed, slowContactsComponent.SprintSpeedModifier);
-            remove = false;
         }
 
         args.ModifySpeed(walkSpeed, sprintSpeed);
-
-        // no longer colliding with anything
-        if (remove)
-            _toRemove.Add(uid);
     }
 
     private void OnEntityExit(EntityUid uid, SlowContactsComponent component, ref EndCollideEvent args)
     {
         var otherUid = args.OtherFixture.Body.Owner;
-        if (HasComp<MovementSpeedModifierComponent>(otherUid))
-            _toUpdate.Add(otherUid);
+
+        if (!TryComp<SlowedByContactComponent>(otherUid, out var slowed))
+        {
+            return;
+        }
+
+        slowed.Intersecting.Remove(uid);
     }
 
     private void OnEntityEnter(EntityUid uid, SlowContactsComponent component, ref StartCollideEvent args)
     {
         var otherUid = args.OtherFixture.Body.Owner;
-        if (!HasComp<MovementSpeedModifierComponent>(otherUid))
+        if (!HasComp<MovementSpeedModifierComponent>(otherUid) ||
+            !TryComp<SlowedByContactComponent>(otherUid, out var slowed))
```

Comment:
The `HasComp<MovementSpeedModifierComponent>` shouldn't be neccesariy anymore now that this uses  `TryComp<SlowedByContactComponent>` instead of `EnsureComp<>`,